### DISCUSSION TITLE
How to best disable transitive package dependencies in SDL2?

### DISCUSSION THREAD

**Discussion by ahrbe1** [microsoft/vcpkg#35306] [Q&A] [answered]
I'd like to ask for some advice on how to deal with an issue I bumped into while trying to use vcpkg for my project. 

My project needs the SDL2 libraries directly, however, it appears I cannot easily override the features since these libraries depend on each other. 

For example, I want "sdl2[x11]", but all the sdl2-* libraries depend on "sdl2". So I tried to use `"default-features": false` in my vcpkg.json file, but I still see it pulling in "sdl2[base,core,ibus,wayland,x11]".

My current vcpkg.json dependencies list looks like this:

```json
    "dependencies": [
        {
            "name": "sdl2",
            "features": ["x11"],
            "default-features": false
        },
        {
            "name": "dbus",
            "platform": "linux",
            "default-features": false,
            "features": [],
            "$comment": "transitive dependency of sdl2 on linux"
        },
        {
            "name": "sdl2-image",
            "default-features": false,
            "features": ["tiff", "libjpeg-turbo"]
        },
        {
            "name": "sdl2-mixer",
            "default-features": false,
            "features": ["libflac", "mpg123"]
        },
        {
            "name": "sdl2-net",
            "default-features": false,
            "features": []
        },
        {
            "name": "sdl2-gfx",
            "default-features": false,
            "features": []
        }
    ]
```

Well, that's not enough apparently, and I still see sdl2 pulling in wayland.

So as a workaround, I patched all of the sdl2-* vcpkg.json port files to remove default features, e.g.:

```diff
   "dependencies": [
-    "sdl2",
+    {
+      "name": "sdl2",
+      "default-features": false
+    },
```

In summary, I guess I found it surprising that `"default-features": false` does not actually disable any default features of transitive dependencies, even though those were directly listed in my vcpkg.json file as such. Ok. So...

The question I have is, "is my workaround really the best way to deal with this situation, or is there a better way?". E.g., do I need to go about all the business of maintaining my own fork of vcpkg that patches these few port files, or is there a more desirable way to disable these dependencies?

**Comment by Osyotr** [comment, marked as answer]
> is my workaround really the best way to deal with this situation, or is there a better way?

Yes. When a package manifest contains `"dependencies": [ "sdl2" ]` it means "I depend on `sdl2` AND its default features". If these features are not required for a successful build, `"default-features": false` should be added.

> do I need to go about all the business of maintaining my own fork of vcpkg that patches these few port files, or is there a more desirable way to disable these dependencies?

You can open PR with these changes in vcpkg repo. (If you maintain your own fork of vcpkg just for these patches, consider using overlay-ports feature or registries)

**Reply by ahrbe1** [reply]
OK. Yeah I wasn't entirely sure if this is something that was intended as-is, or should be fixed in the official repo. But it seems to me that users can't disable those dependencies without some kind of change to the official repo. I'll consider submitting a PR. Thanks.
